File: diameter_of_binary_tree/Diameter_Of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Diameter_Of_Binary_Tree {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }
        
        int diameter = 0;
        dfs(root, diameter);
        return diameter;
    }
    
private:
    int dfs(TreeNode* root, int& diameter) {
        if (root == NULL) {
            return 0;
        }
        
        int left = dfs(root->left, diameter);
        int right = dfs(root->right, diameter);
        diameter = max(diameter, left + right);
        return max(left, right) + 1;
    }
};

File: diameter_of_binary_tree/DiameterOfBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class DiameterOfBinaryTree {
    public int diameterOfBinaryTree(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        int[] ans = new int[1];
        diameterOfBinaryTree(root, ans);
        return ans[0];
    }
    
    private int diameterOfBinaryTree(TreeNode root, int[] ans) {
        if (root == null) {
            return 0;
        }
        
        int left = diameterOfBinaryTree(root.left, ans);
        int right = diameterOfBinaryTree(root.right, ans);
        ans[0] = Math.max(ans[0], left + right);
        
        return Math.max(left, right) + 1;
    }
}
